client/src/main/java/org/asynchttpclient/netty/handler/StreamedResponsePublisher.java

```diff
@@ -42,16 +42,15 @@ public StreamedResponsePublisher(EventExecutor executor, ChannelManager channelM
     protected void cancelled() {
         logger.debug("Subscriber cancelled, ignoring the rest of the body");
 
-        // The subscriber cancelled early, we need to drain the remaining elements from the stream
-        channelManager.drainChannelAndOffer(channel, future);
-        channel.pipeline().remove(StreamedResponsePublisher.class);
-
         try {
             future.done();
         } catch (Exception t) {
             // Never propagate exception once we know we are done.
             logger.debug(t.getMessage(), t);
         }
+
+        // The subscriber cancelled early - this channel is dead and should be closed.
+        channelManager.closeChannel(channel);
```

Comment:
Any reason why you moved this after completing the future?
I think it should be moved back where it was first.

Reply:
Actually yes. There are other places that read the channel which check the future status.
In the previous version, it was possible that channelManager.closeChannel was invoked, the scheduler switches over to another thread, and now that other thread tries to read from the channel that is currently being closed (because its future.isDone check returned false)

Pushing the future.done() further up will avoid that potential threading issue.